rfcs/006 - Replicate changes between staging and prod RFC.md

```diff
@@ -0,0 +1,87 @@
+
+## Alternatives Considered
+
+There are several types of solutions to replicate staging infrastructure changes in production.
+
+### 1. Dump a database and import the dump into another database
+
+An `export` route in the engine implemented using `pgdump` with logic to attempt to make it generic like excluding `aws_account` from the dump and so on. The dump is then intended to be passed into an `import` route for another database. `import` was deprecated for the time being due to the sheer amount of special-casing, which compounded per module, that had to be built within the functionality of export and import.
+
+### 2. Sync a database to a different cloud account than the one it is connected to
+
+An IaSQL function that takes the alias of another database owned by the same user as input. `iasql_import(‘staging’)` will return the same data structure as `iasql_commit` or `iasql_preview`. This function can reuse parts of the mappers and existing IaSQL functions to run diffs across two different databases or accounts. There could have an optional `dry_run` parameter or another function to preview the changes: `iasql_preview_import(‘staging’)` or `iasql_import(‘staging’, true)`.
+
+The steps to replicate recent database/account A changes in database/account B when calling `iasql_import('A')` from within database B:
+1) Regular sync database A to cloud account A
+2) Regular sync database B to cloud account B
+3) New import sync functionality where database B is pointed to cloud account A directly
+
+The new import sync functionality described in #3 would expand the mapper definitions to include a new `import` function and laxer equality function for each module. Capacity differences between production and staging, and AWS-provided cloud IDs would make automatically identifying "equivalent" records from one cloud to fill in for the other very difficult, if not impossible.
+
+### 3. Interactive `git rebase` style sync
+
+Similar to option 2, but when conflict arises prompt the user to handle the merge conflicts between the two cloud accounts or databases interactively. The PostgreSQL REPL doesn't allow interactive input so that would have to happen using a two-phase process via an interim table or the dashboard. This option requires a lot of resources to implement out of the gate. The current proposal and this workflow are compatible, so if this proves to be important to users we can eventually prioritize it.
+
+### 4. Use existing PostgreSQL data synchronization tools
+
+There is a breadth of tools to keep two databases in sync. However, they are not enough to satisfy our requirements for this problem because the data should not be copied over exactly as-is. We have cloud-generated IDs as columns in a lot of tables that are specific to a resource in each account like an EC2 instance id.
```

Comment:
I some kind of "cross-database equality check for entities" would help us. I don't know of any yet, but we can think about it.

Reply:
I'm not sure I understand what you mean here or how it would help use sidestep what is highlighted

Reply:
I mean we need a way to check if two entities are the same thing, but in two different environments (in staging vs in production).
Currently we use `@cloudId` which won't work across accounts, but maybe there's a way? Like tracking the resources and self-assigning them IDs. Anyway this needs more thought.

Reply:
I'm pretty sure it's impossible to do that for some of the reasons laid out in here. Not only are some of them ambiguous (how to distinguish multiple EC2 instances with no `Name` tag) but some may be conceptually identical for the user and how they're using it, but actually different between environments for cost reasons.